### PR TITLE
configure delayed job adapter in non-development environments

### PR DESCRIPTION
**Pull request by davidsbailey**
Per discussion with infra on 9/26, the plan of record is to use Active Job to prevent any long-running requests to openai from tying up our puma processes, barring the discovery of any blockers with this approach. This also lays the groundwork for other features to move off of cronjobs as a mechanism for executing delayed tasks such as sending emails.

Done in this PR:
* installed and configured the delayed_job queue adapter per https://github.com/collectiveidea/delayed_job/tree/v4.1.11#installation
  * a28b5b80c712b94b7bbfdd8a8f47877da6287250: generated by `bundle add delayed_job_active_record`
  * 0b5c0eb07baa5296a1911480b93dc99f676812bb and c64f2b4f32e60c1ddee70f1c0739ffc3da8aaf13: generated by `rails generate delayed_job:active_record ; rake db:migrate`
* 41a98c1c77c649c8de53a9f2e2e214738f09baaf use delayed_job in all non-development environments
   * development environment will still use `:async` adapter, per explicit config in this commit
   *  unit tests will still use the `:test` adapter, per https://guides.rubyonrails.org/testing.html#testing-jobs
* 1fd334e357eef7b473110d9b1411bd66cc385ee0 on all daemon machines, start or restart a single delayed_job worker during rake build

Other background reading:
* Active Job Basics: https://guides.rubyonrails.org/active_job_basics.html
* Active Job adapters: https://api.rubyonrails.org/v7.0.8/classes/ActiveJob/QueueAdapters.html

## Testing story

* verified locally that `bin/delayed_job restart` either starts 1 worker (if none is running), or restarts 1 worker (if one is running)

* pushed this branch to an [adhoc](https://adhoc-delayed-job-adapter-studio.cdn-code.org), verified that the worker process was running, and verified that a job run from dashboard console successfully runs using delayed job:
```
[adhoc] dashboard > SendPlaceholderEmailJob.perform_later(User.first)
Enqueued SendPlaceholderEmailJob (Job ID: 127e33a3-9859-4edc-8e67-544cef587fe2) to DelayedJob(default) with arguments: #<GlobalID:0x00005594484ddc80 @uri=#<URI::GID gid://dashboard/User/1>>

ubuntu@adhoc-delayed-job-adapter:~/adhoc$ grep 'Performed SendPlaceholderEmailJob'  /var/log/syslog
Sep 28 05:18:15 adhoc-delayed-job-adapter dashboard[23630]: Performed SendPlaceholderEmailJob (Job ID: 127e33a3-9859-4edc-8e67-544cef587fe2) from DelayedJob(default) in 56.5ms
```

## Deployment strategy

- [ ] merge https://github.com/code-dot-org/code-dot-org/pull/53975/ , and wait for it to reach production
- [ ] merge this PR, and wait for it to reach production
- [ ] on production-console, verify I can send myself email via `SendPlaceholderEmailJob.perform_later`, and verify it was executed by the delayed job worker 

### number of workers

per yesterday's discussion, Infra team is tasked with figuring out how to run n = 10 or so worker threads in production. I left some comments inline about what I've learned so far on how to do this.

### request latency

by default, each worker is configured to poll the database every 5 seconds. Latency could be reduced by decreasing the `sleep_delay` param to 1 second (for example).

### CODE REVIEW

File: lib/rake/build.rake
Comment (davidsbailey) [inline lib/rake/build.rake:114]: Starting multiple jobs is easy to do if the number of jobs remains fixed:
```
Dave-MBP:~/src/cdo/dashboard (delayed-job-adapter)$ bin/delayed_job -n 2 restart
delayed_job.0: warning: no instances running. Starting...
delayed_job.0: process with pid 66780 started.
delayed_job.1: warning: no instances running. Starting...
delayed_job.1: process with pid 66782 started.
Dave-MBP:~/src/cdo/dashboard (delayed-job-adapter)$ ps -ef | grep delayed | grep -v grep
  501 66780     1   0  4:23PM ??         0:01.39 delayed_job.0
  501 66782     1   0  4:23PM ??         0:01.38 delayed_job.1
```
and it is easy to increase this number:
```
Dave-MBP:~/src/cdo/dashboard (delayed-job-adapter)$ bin/delayed_job -n 4 restart
delayed_job.0: trying to stop process with pid 66780 sending TERM and waiting 20s ...
delayed_job.0: process with pid 66780 successfully stopped.
delayed_job.0: process with pid 67006 started.
delayed_job.1: trying to stop process with pid 66782 sending TERM and waiting 20s ...
delayed_job.1: process with pid 66782 successfully stopped.
delayed_job.1: process with pid 67023 started.
delayed_job.2: warning: no instances running. Starting...
delayed_job.2: process with pid 67025 started.
delayed_job.3: warning: no instances running. Starting...
delayed_job.3: process with pid 67027 started.
Dave-MBP:~/src/cdo/dashboard (delayed-job-adapter)$ ps -ef | grep delayed | grep -v grep
  501 67006     1   0  4:25PM ??         0:01.33 delayed_job.0
  501 67023     1   0  4:25PM ??         0:00.47 delayed_job.1
  501 67025     1   0  4:25PM ??         0:00.36 delayed_job.2
  501 67027     1   0  4:25PM ??         0:00.36 delayed_job.3
```
but this naive approach leads to unacceptable behavior if the number of workers is later lowered:
```
Dave-MBP:~/src/cdo/dashboard (delayed-job-adapter)$ bin/delayed_job -n 3 restart
delayed_job.0: trying to stop process with pid 67006 sending TERM and waiting 20s ...
delayed_job.0: process with pid 67006 successfully stopped.
delayed_job.0: process with pid 67220 started.
delayed_job.1: trying to stop process with pid 67023 sending TERM and waiting 20s ...
delayed_job.1: process with pid 67023 successfully stopped.
delayed_job.1: process with pid 67238 started.
delayed_job.2: trying to stop process with pid 67025 sending TERM and waiting 20s ...
delayed_job.2: process with pid 67025 successfully stopped.
delayed_job.2: process with pid 67253 started.
Dave-MBP:~/src/cdo/dashboard (delayed-job-adapter)$ ps -ef | grep delayed | grep -v grep
  501 67027     1   0  4:25PM ??         0:12.15 delayed_job.3
  501 67220     1   0  4:26PM ??         0:02.32 delayed_job.0
  501 67238     1   0  4:26PM ??         0:01.32 delayed_job.1
  501 67253     1   0  4:26PM ??         0:00.47 delayed_job.2
```
not only is delayed_job.3 still running, but the PID is unchanged which means that it will be running old code indefinitely.

Comment (davidsbailey) [reply lib/rake/build.rake:114]: full command line options can be found here: https://github.com/collectiveidea/delayed_job/blob/92fb435131d05b8a8715ac52acaacd9cde89619c/lib/delayed/command.rb#L28

File: dashboard/bin/delayed_job
Comment (cat5inthecradle) [inline dashboard/bin/delayed_job:2]:
```suggestion

# This script sets up and runs a Delayed Job worker in the background.
```

File: lib/rake/build.rake
Comment (cat5inthecradle) [reply lib/rake/build.rake:114]: Is this script only ever executed on daemon servers?

I'm wondering if we should split this and "bin/delayed_job" out of this PR. The rest just sets us up to create jobs and add them to the database queue and I approve of those changes. My comments and questions are only in this build command change, and the "bin/delayed_job" script.

Comment (davidsbailey) [reply lib/rake/build.rake:114]: > Is this script only ever executed on daemon servers?

yes, it's a long if clause starting here: https://github.com/code-dot-org/code-dot-org/blob/1fd334e357eef7b473110d9b1411bd66cc385ee0/lib/rake/build.rake#L58

> I'm wondering if we should split this and "bin/delayed_job" out of this PR. The rest just sets us up to create jobs and add them to the database queue and I approve of those changes. My comments and questions are only in this build command change, and the "bin/delayed_job" script.

Sure: https://github.com/code-dot-org/code-dot-org/pull/54021 Please keep in mind that bin/delayed_job is generated by `rails generate delayed_job:active_record` and its usage is referenced https://github.com/collectiveidea/delayed_job/blob/master/README.md, so we wouldn't want to change its behavior too drastically, though could see wanting to add comments or wrap it with certain protections in the prod environment or something like that.

File: dashboard/bin/delayed_job
Comment (davidsbailey) [reply dashboard/bin/delayed_job:2]: 4aff5d30c30f310189b2207f6c46a886c7fb5efb

File: lib/rake/build.rake
Comment (davidsbailey) [reply lib/rake/build.rake:114]: > I'm wondering if we should split this and "bin/delayed_job" out of this PR. The rest just sets us up to create jobs and add them to the database queue and I approve of those changes. My comments and questions are only in this build command change, and the "bin/delayed_job" script.
 
@cat5inthecradle any remaining concerns now that those things have been split out? if not, it would be great to get your approval on this and https://github.com/code-dot-org/code-dot-org/pull/53975 so that we can start chipping away at the stack of PRs here.

Comment (cat5inthecradle) [reply lib/rake/build.rake:114]: This PR looks good to me, and I'm fine merging the other one into this one, then this one into staging. I'm dropping a note/question on the other PR though.

**Review by cat5inthecradle — APPROVED** [review]

File: lib/rake/build.rake
Comment (davidsbailey) [reply lib/rake/build.rake:114]: Woohoo, thank you for the careful consideration here Darin!

**Comment by davidsbailey** [comment]
after merging to staging, I confirmed that the worker is running and that delayed jobs can be enqueued from rails console and performed by the worker:
```
ubuntu@staging:~/staging$ bin/dashboard-console
[staging] dashboard > u = User.find_by_email('dave@code.org')
[staging] dashboard > SendPlaceholderEmailJob.perform_later(u)
  TRANSACTION (9.7ms)  BEGIN
  Delayed::Backend::ActiveRecord::Job Create (0.9ms)  INSERT INTO `delayed_jobs` (`handler`, `run_at`, `queue`, `created_at`, `updated_at`) VALUES ('--- !ruby/object:ActiveJob::QueueAdapters::DelayedJobAdapter::JobWrapper\njob_data:\n  job_class: SendPlaceholderEmailJob\n  job_id: 2f6df2ed-d530-4167-b099-9894897687df\n  provider_job_id:\n  queue_name: default\n  priority:\n  arguments:\n  - _aj_globalid: gid://dashboard/User/75\n  executions: 0\n  exception_executions: {}\n  locale: en-US\n  timezone: UTC\n  enqueued_at: \'2023-10-06T23:25:28Z\'\n', '2023-10-06 23:25:28', 'default', '2023-10-06 23:25:28.772553', '2023-10-06 23:25:28.772553')
  TRANSACTION (0.3ms)  COMMIT
Enqueued SendPlaceholderEmailJob (Job ID: 2f6df2ed-d530-4167-b099-9894897687df) to DelayedJob(default) with arguments: #<GlobalID:0x00005609ce8d0990 @uri=#<URI::GID gid://dashboard/User/75>>
```

```
ubuntu@staging:~/staging$ sudo tail -f /var/log/syslog | grep Job
...
Oct  6 23:25:28 staging dashboard[4091032]: Enqueued SendPlaceholderEmailJob (Job ID: 2f6df2ed-d530-4167-b099-9894897687df) to DelayedJob(default) with arguments: #<GlobalID:0x00005609ce8dbf70 @uri=#<URI::GID gid://dashboard/User/75>>
Oct  6 23:25:30 staging dashboard[4079213]: Delayed::Backend::ActiveRecord::Job Update All (0.6ms)  UPDATE `delayed_jobs` SET `delayed_jobs`.`locked_at` = '2023-10-06 23:25:30', `delayed_jobs`.`locked_by` = 'delayed_job host:staging pid:4079213' WHERE (((run_at <= '2023-10-06 23:25:30.974408' AND (locked_at IS NULL OR locked_at < '2023-10-06 19:25:30.974428')) OR locked_by = 'delayed_job host:staging pid:4079213') AND failed_at IS NULL) ORDER BY priority ASC, run_at ASC LIMIT 1
Oct  6 23:25:30 staging dashboard[4079213]: Delayed::Backend::ActiveRecord::Job Load (0.4ms)  SELECT `delayed_jobs`.* FROM `delayed_jobs` WHERE `delayed_jobs`.`locked_at` = '2023-10-06 23:25:30' AND `delayed_jobs`.`locked_by` = 'delayed_job host:staging pid:4079213' AND `delayed_jobs`.`failed_at` IS NULL ORDER BY `delayed_jobs`.`id` ASC LIMIT 1
Oct  6 23:25:30 staging dashboard[4079213]: 2023-10-06T23:25:30+0000: [Worker(delayed_job host:staging pid:4079213)] Job SendPlaceholderEmailJob [2f6df2ed-d530-4167-b099-9894897687df] from DelayedJob(default) with arguments: [{"_aj_globalid"=>"gid://dashboard/User/75"}] (id=1) (queue=default) RUNNING
Oct  6 23:25:31 staging dashboard[4079213]: Performing SendPlaceholderEmailJob (Job ID: 2f6df2ed-d530-4167-b099-9894897687df) from DelayedJob(default) enqueued at 2023-10-06T23:25:28Z with arguments: #<GlobalID:0x000055ac2c21b870 @uri=#<URI::GID gid://dashboard/User/75>>
Oct  6 23:25:31 staging dashboard[4079213]: Performed SendPlaceholderEmailJob (Job ID: 2f6df2ed-d530-4167-b099-9894897687df) from DelayedJob(default) in 54.08ms
```

**Comment by davidsbailey** [comment]
On the test machine, I can enqueue a job:
```
[test] dashboard > SendPlaceholderEmailJob.perform_later(u)
Enqueued SendPlaceholderEmailJob (Job ID: 191d7021-4d6a-4252-a26d-6eb113c071b3) to DelayedJob(default) with arguments: #<GlobalID:0x00007f26eba15130 @uri=#<URI::GID gid://dashboard/User/507370>>
```

and I can see (in a different log file) that the job is completed:
```
ubuntu@test:~/test$ grep SendPlaceholderEmailJob dashboard/log/test.log
[ActiveJob] Enqueued SendPlaceholderEmailJob (Job ID: 191d7021-4d6a-4252-a26d-6eb113c071b3) to DelayedJob(default) with arguments: #<GlobalID:0x00007f26eba14640 @uri=#<URI::GID gid://dashboard/User/507370>>
2023-10-07T03:10:02+0000: [Worker(delayed_job host:test pid:283402)] Job SendPlaceholderEmailJob [191d7021-4d6a-4252-a26d-6eb113c071b3] from DelayedJob(default) with arguments: [{"_aj_globalid"=>"gid://dashboard/User/507370"}] (id=3) (queue=default) RUNNING
[ActiveJob] [SendPlaceholderEmailJob] [191d7021-4d6a-4252-a26d-6eb113c071b3] Performing SendPlaceholderEmailJob (Job ID: 191d7021-4d6a-4252-a26d-6eb113c071b3) from DelayedJob(default) enqueued at 2023-10-07T03:09:59Z with arguments: #<GlobalID:0x000055f2d7920520 @uri=#<URI::GID gid://dashboard/User/507370>>
[ActiveJob] [SendPlaceholderEmailJob] [191d7021-4d6a-4252-a26d-6eb113c071b3]   Rendered placeholder_mailer/placeholder_email.haml (Duration: 0.7ms | Allocations: 42)
[ActiveJob] [SendPlaceholderEmailJob] [191d7021-4d6a-4252-a26d-6eb113c071b3] Delivered mail 6520cc0a66ae5_4530a44349465d@test.mail (3.2ms)
[ActiveJob] [SendPlaceholderEmailJob] [191d7021-4d6a-4252-a26d-6eb113c071b3] Performed SendPlaceholderEmailJob (Job ID: 191d7021-4d6a-4252-a26d-6eb113c071b3) from DelayedJob(default) in 13.87ms
2023-10-07T03:10:02+0000: [Worker(delayed_job host:test pid:283402)] Job SendPlaceholderEmailJob [191d7021-4d6a-4252-a26d-6eb113c071b3] from DelayedJob(default) with arguments: [{"_aj_globalid"=>"gid://dashboard/User/507370"}] (id=3) (queue=default) COMPLETED after 0.0234
```

I don't actually receive the email on test like I did on staging, but this is explained by test being configured differently for email: https://github.com/code-dot-org/code-dot-org/blob/6ad08d8c28f4b459420b53c5e4c9d970796d0f2a/dashboard/config/environments/test.rb#L56-L59

**Comment by davidsbailey** [comment]
Everything seems to be working on production:

email enqueued via rails console on production-**console**:
```
ubuntu@production-console:~/production$ bin/dashboard-console
[production] dashboard > u = User.find_by_email('dave@code.org')
[production] dashboard > SendPlaceholderEmailJob.perform_later(u)
Enqueued SendPlaceholderEmailJob (Job ID: 6592fa18-570c-4b0a-afb5-3a6774907c65) to DelayedJob(default) with arguments: #<GlobalID:0x00005642f1116158 @uri=#<URI::GID gid://dashboard/User/56232731>>
```

job performed on production-**daemon**:
```
ubuntu@production-daemon:~/production$ grep SendPlaceholderEmailJob /var/log/syslog
Oct  9 22:36:51 production-daemon dashboard[21145]: 2023-10-09T22:36:51+0000: [Worker(delayed_job host:production-daemon pid:21145)] Job SendPlaceholderEmailJob [6592fa18-570c-4b0a-afb5-3a6774907c65] from DelayedJob(default) with arguments: [{"_aj_globalid"=>"gid://dashboard/User/56232731"}] (id=1) (queue=default) RUNNING
Oct  9 22:36:51 production-daemon dashboard[21145]: Performing SendPlaceholderEmailJob (Job ID: 6592fa18-570c-4b0a-afb5-3a6774907c65) from DelayedJob(default) enqueued at 2023-10-09T22:36:46Z with arguments: #<GlobalID:0x00005636da118e58 @uri=#<URI::GID gid://dashboard/User/56232731>>
Oct  9 22:36:51 production-daemon dashboard[21145]: Performed SendPlaceholderEmailJob (Job ID: 6592fa18-570c-4b0a-afb5-3a6774907c65) from DelayedJob(default) in 33.53ms
Oct  9 22:36:51 production-daemon dashboard[21145]: 2023-10-09T22:36:51+0000: [Worker(delayed_job host:production-daemon pid:21145)] Job SendPlaceholderEmailJob [6592fa18-570c-4b0a-afb5-3a6774907c65] from DelayedJob(default) with arguments: [{"_aj_globalid"=>"gid://dashboard/User/56232731"}] (id=1) (queue=default) COMPLETED after 0.0415
```
Email received in my inbox:
Screenshot 2023-10-09 at 3 40 31 PM

**Pull request merged**
